README.md
WSJTX to MQTT
=============

[![Available on Hackage][badge-hackage]][hackage]
[![License BSD3][badge-license]][license]
[![Build Status][badge-travis]][travis]
[![AppVeyor][badge-appveyor]][appveyor]
[![Windows Binary Release][badge-github-releases]][github-releases]

This is a bridge between WSJTX amd MQTT.
One end listens for incoming UDP packages from WSJTX
the other end works as a MQTT client.

[travis]: https://travis-ci.org/MarcFontaine/wsjtx-to-mqtt
[badge-travis]: https://img.shields.io/travis/MarcFontaine/wsjtx-to-mqtt.svg?label=Linux%20build
[appveyor]: https://ci.appveyor.com/project/MarcFontaine/wsjtx-to-mqtt/branch/master
[badge-appveyor]: https://img.shields.io/appveyor/ci/MarcFontaine/wsjtx-to-mqtt.svg?label=Windows%20build
[badge-github-releases]: https://img.shields.io/github/release/MarcFontaine/wsjtx-to-mqtt.svg?label=Windows%20Binary
[github-releases]: https://github.com/MarcFontaine/wsjtx-to-mqtt/releases
[badge-license]: https://img.shields.io/badge/license-BSD3-green.svg
[license]: https://github.com/MarcFontaine/wsjtx-udp/blob/master/LICENSE
[hackage]: https://hackage.haskell.org/package/wsjtx-to-mqtt
[badge-hackage]: https://img.shields.io/hackage/v/wsjtx-to-mqtt.svg


CHANGELOG.md
## 0.1.4.1
* add Changelog
* update CI


src/MQTT.hs
{-# LANGUAGE OverloadedStrings #-}
module MQTT
where

import Data.Text as Text (Text, intercalate)
import Data.Text.Encoding (encodeUtf8)
import qualified Data.ByteString.Char8 as BSC
import Data.ByteString.Builder
import Data.ByteString.Lazy as BSL

import Control.Exception.Base (bracket)
import Control.Concurrent (threadDelay, forkIO)
import Control.Concurrent.Chan (newChan, readChan, writeChan)
import Control.Monad (forM_, forever, void)
import Network.Socket hiding (send, recv, recvFrom)
import Network.Socket.ByteString.Lazy (sendAll, recv)

data PublishConfig = PublishConfig
  {
     _pcUsername :: Maybe String
   , _pcPassword :: Maybe String
   , _pcHost     :: String
   , _pcPort     :: String
   , _pcClientID :: String
  } deriving Show

defaultPublishConfig :: PublishConfig
defaultPublishConfig = PublishConfig
  {
     _pcUsername = Nothing
   , _pcPassword = Nothing
   , _pcHost     = "127.0.0.1"
   , _pcPort     = "1883"
   , _pcClientID = "hspub01"
  }

data Msg = Msg [Text] !BSL.ByteString
  deriving Show

withMqttPublish :: PublishConfig -> ((Msg -> IO ()) -> IO ()) -> IO ()
withMqttPublish config callback = withSocketsDo $ do
    addr <- resolve (_pcHost config) (_pcPort config)
    channel <- newChan
    void $ forkIO $ bracket (open addr) close $ \sock -> do
        sendAll sock myConnectMsg
        forever $ do
          readChan channel >>= return . publishMsg >>= sendAll sock 
          -- TODO : read any data from the socket
    callback (writeChan channel)
  where
    resolve host port = do
        let hints = defaultHints { addrSocketType = Stream }
        addr:_ <- getAddrInfo (Just hints) (Just host) (Just port)
        return addr
    open addr = do
        sock <- socket (addrFamily addr) (addrSocketType addr) (addrProtocol addr)
        connect sock $ addrAddress addr
        return sock
    myConnectMsg = connectMsg (BSC.pack $ _pcClientID config)

testPublish :: IO ()
testPublish = withMqttPublish defaultPublishConfig $ \publish -> do
    publish $ Msg ["hamradio","topic"] "msg1"
    publish $ Msg ["hamradio","topic"] "msg2"
    publish $ Msg ["hamradio","topic3"] "msg3"
    threadDelay 5000000    

sendMqtt :: [BSL.ByteString] -> IO ()    
sendMqtt packets = withSocketsDo $ do
    addr <- resolve "127.0.0.1" "1883"
    bracket (open addr) close talk
  where
    resolve host port = do
        let hints = defaultHints { addrSocketType = Stream }
        addr:_ <- getAddrInfo (Just hints) (Just host) (Just port)
        return addr
    open addr = do
        sock <- socket (addrFamily addr) (addrSocketType addr) (addrProtocol addr)
        connect sock $ addrAddress addr
        return sock
    talk sock = do
        forM_ packets $ sendAll sock
        _msg <- recv sock 1024
        Prelude.putStr "Received: "
--        BSC.putStrLn msg
        threadDelay 5000000
                  
connectMsg :: BSC.ByteString -> BSL.ByteString
connectMsg clientId = toLazyByteString $
       connectHeader
    `mappend` remainingLength (
           protocolName
        `mappend` protocolLevel
        `mappend` connectFlags
        `mappend` keepAlive
        `mappend` lenBS clientId
        )  
    where
        connectHeader = word8 0x10      
        protocolName  = lenBS "MQIsdp"
        protocolLevel = word8 0x03
        connectFlags  = word8 0x02
        keepAlive     = word16BE 0

publishMsg :: Msg -> BSL.ByteString
publishMsg (Msg topic message) = toLazyByteString $
    ( pubHeader
      `mappend` remainingLength (
           (lenBS $ encodeUtf8 $ Text.intercalate "/" topic)
        `mappend` identifier
        `mappend` lazyByteString message ))
    where
        pubHeader = word8 0x30
        identifier = mempty 

lenBS :: BSC.ByteString -> Builder
lenBS bs
    = ( word16BE $ fromIntegral $ BSC.length bs) `mappend` byteString bs
    
remainingLength :: Builder -> Builder
remainingLength c =
     word8 len -- variable length
  `mappend` lazyByteString body
  where
    body = toLazyByteString c
    len = fromIntegral $ BSL.length body


src/Mqtt.hs
{-# LANGUAGE OverloadedStrings #-}
module Mqtt
where

import Data.Text as Text (Text, intercalate)
import Data.Text.Encoding (encodeUtf8)
import qualified Data.ByteString.Char8 as BSC
import Data.ByteString.Builder
import Data.ByteString.Lazy as BSL

import Control.Exception.Base (bracket)
import Control.Concurrent (threadDelay, forkIO)
import Control.Concurrent.Chan (newChan, readChan, writeChan)
import Control.Monad (forM_, forever, void)
import Network.Socket hiding (send, recv, recvFrom)
import Network.Socket.ByteString.Lazy (sendAll, recv)

data PublishConfig = PublishConfig
  {
     _pcUsername :: Maybe String
   , _pcPassword :: Maybe String
   , _pcHost     :: String
   , _pcPort     :: String
   , _pcClientID :: String
  } deriving Show

defaultPublishConfig :: PublishConfig
defaultPublishConfig = PublishConfig
  {
     _pcUsername = Nothing
   , _pcPassword = Nothing
   , _pcHost     = "127.0.0.1"
   , _pcPort     = "1883"
   , _pcClientID = "hspub01"
  }

data Msg = Msg [Text] !BSL.ByteString
  deriving Show

withMqttPublish :: PublishConfig -> ((Msg -> IO ()) -> IO ()) -> IO ()
withMqttPublish config callback = withSocketsDo $ do
    addr <- resolve (_pcHost config) (_pcPort config)
    channel <- newChan
    void $ forkIO $ bracket (open addr) close $ \sock -> do
        sendAll sock myConnectMsg
        forever $ do
          readChan channel >>= return . publishMsg >>= sendAll sock 
          -- TODO : read any data from the socket
    callback (writeChan channel)
  where
    resolve host port = do
        let hints = defaultHints { addrSocketType = Stream }
        addr:_ <- getAddrInfo (Just hints) (Just host) (Just port)
        return addr
    open addr = do
        sock <- socket (addrFamily addr) (addrSocketType addr) (addrProtocol addr)
        connect sock $ addrAddress addr
        return sock
    myConnectMsg = connectMsg (BSC.pack $ _pcClientID config)

testPublish :: IO ()
testPublish = withMqttPublish defaultPublishConfig $ \publish -> do
    publish $ Msg ["hamradio","topic"] "msg1"
    publish $ Msg ["hamradio","topic"] "msg2"
    publish $ Msg ["hamradio","topic3"] "msg3"
    threadDelay 5000000    

sendMqtt :: [BSL.ByteString] -> IO ()    
sendMqtt packets = withSocketsDo $ do
    addr <- resolve "127.0.0.1" "1883"
    bracket (open addr) close talk
  where
    resolve host port = do
        let hints = defaultHints { addrSocketType = Stream }
        addr:_ <- getAddrInfo (Just hints) (Just host) (Just port)
        return addr
    open addr = do
        sock <- socket (addrFamily addr) (addrSocketType addr) (addrProtocol addr)
        connect sock $ addrAddress addr
        return sock
    talk sock = do
        forM_ packets $ sendAll sock
        _msg <- recv sock 1024
        Prelude.putStr "Received: "
--        BSC.putStrLn msg
        threadDelay 5000000
                  
connectMsg :: BSC.ByteString -> BSL.ByteString
connectMsg clientId = toLazyByteString $
       connectHeader
    <> remainingLength (
           protocolName
        <> protocolLevel
        <> connectFlags
        <> keepAlive
        <> lenBS clientId
        )  
    where
        connectHeader = word8 0x10      
        protocolName  = lenBS "MQIsdp"
        protocolLevel = word8 0x03
        connectFlags  = word8 0x02
        keepAlive     = word16BE 0

publishMsg :: Msg -> BSL.ByteString
publishMsg (Msg topic message) = toLazyByteString $
    ( pubHeader
      <> remainingLength (
           (lenBS $ encodeUtf8 $ Text.intercalate "/" topic)
        <> identifier
        <> lazyByteString message ))
    where
        pubHeader = word8 0x30
        identifier = mempty 

lenBS :: BSC.ByteString -> Builder
lenBS bs
    = ( word16BE $ fromIntegral $ BSC.length bs) <> byteString bs
    
remainingLength :: Builder -> Builder
remainingLength c =
     word8 len -- variable length
  <> lazyByteString body
  where
    body = toLazyByteString c
    len = fromIntegral $ BSL.length body


src/Bridge.hs
----------------------------------------------------------------------------
-- |
-- Module      :  Bridge
-- Copyright   :  (c) Marc Fontaine 2017-2018
-- License     :  BSD3
--
-- Maintainer  :  Marc.Fontaine@gmx.de
-- Stability   :  experimental
-- Portability :  GHC-only
--
{-# LANGUAGE OverloadedStrings #-}
module Bridge where

import           Control.Concurrent (forkIO, threadDelay)
import           Control.Concurrent.MVar
import           Control.Monad
import           Data.ByteString.Lazy.Char8 as BSLC (unpack)
import           Data.Text as Text

import           MQTT
import qualified WSJTX.UDP.NetworkMessage as WSJTX
import           WSJTX.UDP.Server

import           FormatMqttMsg (toMqttMsg)
import           W2MTypes

runBridge :: W2MTypes.Config -> IO ()
runBridge config
    = withWsjtxSocket wsjtxPort
      $ \wsjtxSocket -> withMqttPublish mqttConfig $ \publish -> do
    debugPrint config "starting processes"
    wsjtxState <- newMVar Nothing
    _wsjtxThread <- forkWsjtxServer wsjtxSocket
                      (udpConsumer
                       config
                       publish
                       wsjtxState
                      )
                      
    debugPrint config "starting processes2" 
    case (beacon_interval serverConf) of
        Nothing -> return ()
        Just t -> do
            debugPrint config "starting beacon transmitter"
            forkBeacon t config publish wsjtxState

    _ <- getLine
    return ()
    where
        wsjtxPort = fromInteger $ udp_port $ config_wsjtx config
        serverConf = config_report_broker config
        mqttConfig = MQTT.PublishConfig
               {
                _pcUsername = Nothing
              , _pcPassword = Nothing
              , _pcHost     = "127.0.0.1"
              , _pcPort     = "1883"
              , _pcClientID = "hspub01"
              }

        udpConsumer mainConfig publish wsjtxStatus package = do
            debugPrint config $ "UDP recieved :" ++ show package
            status <- fmap join $ tryReadMVar wsjtxStatus
            case package of
                WSJTX.PDecode msg -> do
                    let (t,m) =  toMqttMsg mainConfig status msg
                    _ <- publish $ Msg t m
                    debugPrint config $ "MQTT send :" ++ BSLC.unpack m

                WSJTX.PStatus s -> modifyMVar_ wsjtxStatus (const $ return $ Just s)
                _  -> return ()

        forkBeacon delay _mainConfig publish _wsjtxStatus
            = void $ forkIO $ forM_ [(1::Int)..] $ \cnt -> do
            let t =  ( head_topics $ config_report_broker config )
                  ++ [ beacon_topic $ config_report_broker config ]
                  ++ [ Text.pack $ show cnt ]
            _ <- publish $ Msg t "ping"
            debugPrint config $ "MQTT send : ping"
            threadDelay $ delay*1000


src/Config.hs
----------------------------------------------------------------------------
-- |
-- Module      :  Config
-- Copyright   :  (c) Marc Fontaine 2017-2018
-- License     :  BSD3
--
-- Maintainer  :  Marc.Fontaine@gmx.de
-- Stability   :  experimental
-- Portability :  GHC-only
--

{-# LANGUAGE QuasiQuotes #-}
module Config where

import           Control.Monad
import qualified Data.Default.Class as Default
import           Data.Maybe
import           Data.Yaml as Yaml
import           System.Environment (lookupEnv)
import           System.Exit (exitFailure)

import           W2MTypes

getConfig :: MainOptions -> IO W2MTypes.Config
getConfig opts
  = if _testDefault opts
       then return Default.def
       else do
  envPath <- lookupEnv "WSJTX_TO_MQTT_CONFIG_PATH"
  let configFile = fromJust (
        (_configPath opts)
        `mplus` envPath
        `mplus` (Just "wsjtx-to-mqtt.yaml")
        )
  when (_debug opts) $ putStrLn $ "config file path :" ++ configFile
  confData <- Yaml.decodeFileEither configFile
  config <- case confData of
    Right x  -> return $ x {config_debug =config_debug x || _debug opts}
    Left err -> do
      putStrLn "Cannot parse config file"
      putStrLn $ show err
      exitFailure
  return config


helpConfig :: IO ()
helpConfig = do
    putStrLn "# possible config file :"
    putStrLn includedConfigStr

includedConfigStr :: String
includedConfig :: W2MTypes.Config
(includedConfigStr,includedConfig) = [quoteConfig|
# some comment
wsjtx:
  udp_port: 2237
protocol_version: W2M-V0.1
reporter:
  info: ! 'rig : FT-840,  Antenna : Vertical'
  callsign: E8TST
  band_config:
    tag: TrackWSJTXBand
  locator: FE43ab
report_broker:
  username: myUserName
  password: secret

#  username: null
#  password: null

  head_topics:
  - hamradio
  - spot
  host: localhost
# host: broker.hivemq.com
# host: test.mosquitto.org
# host: iot.eclipse.org ?

# one ping every 1000 ms
  beacon_interval: 1000
  beacon_topic: test-beacon
enable_report: true
debug: false
|]


src/WsjtxToMqtt.hs
{-# Language DataKinds, OverloadedStrings, ScopedTypeVariables #-}
{-# LANGUAGE RecordWildCards, DeriveGeneric, DuplicateRecordFields #-}
module Main where

import Control.Monad
import System.Exit (exitFailure, exitSuccess)
import Paths_wsjtx_to_mqtt (version)
import qualified Data.ByteString.Lazy.Char8 as BSLC (pack,putStrLn)
import qualified Data.ByteString.Char8 as BS (putStrLn, unpack)
import qualified Data.Text as Text (unpack)
import qualified Options (runSubcommand, subcommand)
import Data.Aeson as Aeson (eitherDecode, encode)
import Data.Yaml as Yaml (encode)
import Data.Version (showVersion)
import Data.Double.Conversion.ByteString (toFixed)
import Network.Socket (withSocketsDo)

import W2MTypes
import Bridge (runBridge,testMqtt)
import Config (getConfig, helpConfig)
import WSJTX.UDP.Server (testDump, replyWithPackages, withWsjtxSocket, forkWsjtxServer)
import WSJTX.UDP.NetworkMessage (Package(..), Decode(..))

main :: IO ()
main = withSocketsDo $ Options.runSubcommand [
   Options.subcommand "showConfig" showConfig
  ,Options.subcommand "forward" forward
  ,Options.subcommand "dumpWsjtx" dumpWsjtx
  ,Options.subcommand "sendToUdp" sendToUDP
  ,Options.subcommand "testMqtt" $ \_ EmptyOptions _ -> testMqtt
  ]

dumpWsjtx :: MainOptions -> DumpOptions -> [String] -> IO ()
dumpWsjtx _ format _ = case format of
    DumpHaskell -> void $ testDump
    DumpText    -> server dumpText
    DumpJSON    -> server $ BSLC.putStrLn . Aeson.encode
    where
        server :: (Package -> IO ()) -> IO ()
        server dump = withWsjtxSocket wsjtxPort $ \sock -> do
                        _threadId <- forkWsjtxServer sock dump
                        void getLine
          
        wsjtxPort = 2237

        dumpText :: Package -> IO ()
        dumpText (PDecode (Decode {..})) = putStrLn $ concat [
            show decode_time
          , pad 4 $ BS.unpack $ toFixed 1 decode_delta_time
          , pad 5 $ show decode_snr
          , pad 5 $ show decode_delta_frequency
          , " ~  "
          , Text.unpack decode_message
          ]
        dumpText _ = return ()
        pad l s = reverse $ take l $ reverse $ "    " ++ s

showConfig :: MainOptions -> EmptyOptions -> [String] -> IO ()
showConfig cmdOpts EmptyOptions _ = do
  putStrLn $ "# Version : " ++ showVersion version
  helpConfig
  putStrLn "---------------" 
  putStrLn "current config:"
  config <- getConfig cmdOpts
  BS.putStrLn $ Yaml.encode config
  exitSuccess

sendToUDP :: MainOptions -> EmptyOptions -> [String] -> IO ()
sendToUDP cmdOpts EmptyOptions strs = do
  config <- getConfig cmdOpts
  let port = fromInteger $ udp_port $ config_wsjtx config
  let decoded = forM strs (Aeson.eitherDecode . BSLC.pack)
  case decoded of
       Right packageList -> replyWithPackages port packageList
       Left msg -> do
         putStrLn "error : cannot decode packages"
         putStrLn $ show (msg , strs)
         exitFailure
  exitSuccess

forward :: MainOptions -> EmptyOptions -> [String] -> IO ()
forward cmdOpts EmptyOptions _args = do
  config <- getConfig cmdOpts
  debugPrint config "starting wsjtx to udp bridge"
  runBridge config


src/FormatMqttMsg.hs
{-# LANGUAGE OverloadedStrings #-}
module FormatMqttMsg where

import           Data.Aeson as Aeson
import qualified Data.ByteString.Lazy as BSL
import           Data.Fixed
import           Data.Text as Text
import           Data.Word

import           W2MTypes
import qualified WSJTX.UDP.NetworkMessage as WSJTX

toMqttMsg ::
     W2MTypes.Config -> Maybe WSJTX.Status -> WSJTX.Decode
  -> ([Text] , BSL.ByteString)
toMqttMsg config status msg
  = (toTopic config status msg report
    ,Aeson.encode report)
  where
    report = toReport config status msg

toReport ::
  W2MTypes.Config -> Maybe WSJTX.Status -> WSJTX.Decode -> Report
toReport config status msg = report
  where
    maybeBand = fmap WSJTX.status_dial_frequency status >>= freqToBand
    bandInfo = case reporter_band_config $ config_reporter config of
       UnknownBand -> "unknown_band"
       FixedBand b  -> b
       TrackWSJTXBand -> case maybeBand of
         Nothing -> "unknown_band"
         Just b  -> b
    reporter = case reporter_callsign $ config_reporter config of
       Nothing -> "unknown_reporter"
       Just cs -> cs
    report = Report {
        message = WSJTX.decode_message msg
       ,mode = maybe  "unknown_mode" WSJTX.status_mode status
       ,band = bandInfo
       ,freq = maybe 0 WSJTX.status_dial_frequency status
       ,time = WSJTX.decode_time msg
       ,snr  = WSJTX.decode_snr msg
       ,delta_time = MkFixed $ round (WSJTX.decode_delta_time msg *100)
       ,delta_frequency = WSJTX.decode_delta_frequency msg
       ,recv_locator = reporter_locator $ config_reporter config
       ,recv_callsign = reporter
       ,recv_info = reporter_info $ config_reporter config
     }

toTopic ::
       W2MTypes.Config -> Maybe WSJTX.Status -> WSJTX.Decode -> Report
   -> [Text]
toTopic config _status _msg report
   =    ( head_topics $ config_report_broker config )
     ++ [
         band report
        ,mode report
        ,"callsign"
        ,recv_callsign report
        ,Text.take 2 $ recv_locator report
        ,if (Text.take 2 $ message report) == "CQ" then "is_cq" else "no_cq"
        -- when
        ,config_protocol_version config
        ]

freqToBand :: Word64 -> Maybe Text
freqToBand f
  = case Prelude.filter isInBand wsjtxBands of
      [(bnd,_)] -> Just bnd
      _         -> Nothing
  where
    overlap = 20000 -- 20kz overlap
    isInBand (_,(s,e)) = s-overlap <= f && f <= e + overlap
    wsjtxBands = [
       ("160m",(  1810000,   2000000))
      ,("80m", (  3500000,   3800000))
      ,("40m", (  7000000,   7200000))
      ,("30m", ( 10100000,  10150000))
      ,("20m", ( 14000000,  14350000))
      ,("17m", ( 18068000,  18168000))
      ,("15m", ( 21000000,  21450000))
      ,("12m", ( 24890000,  24990000))
      ,("10m", ( 28000000,  29700000))
      ,("6m",  ( 50030000,  51000000))
      ,("4m",  ( 70000000,  70500000))
      ]


src/ForwardToMqtt.hs
module Bridge where

import Control.Concurrent.STM
import Control.Concurrent.MVar
import Control.Monad

import Network.MQTT hiding (defaultConfig)
import qualified Network.MQTT as MQTT
import WSJTX.UDP.Server
import qualified WSJTX.UDP.NetworkMessage as WSJTX

import W2MTypes
import FormatMqttMsg
  
runBridge :: W2MTypes.Config -> IO ()
runBridge config = withWsjtxSocket wsjtxPort $ \wsjtxSocket -> do
  cmds <- mkCommands
  pubChan <- newTChanIO
  let serverConf = config_mqtt_client config
      mqttConf = (MQTT.defaultConfig cmds pubChan)
              { cUsername = W2MTypes.username serverConf
              , cPassword = W2MTypes.password serverConf
              , cHost     = host serverConf
              }
  wsjtxState <- newMVar Nothing
  _wsjtxTread <- forkWsjtxServer wsjtxSocket 
                    (udpConsumer
                       config
                       mqttConf
                       wsjtxState
                    )

  terminated <- MQTT.run mqttConf
  print terminated
  where
    wsjtxPort = fromInteger $ udp_port $ config_wsjtx config

    udpConsumer mainConfig mqttConf wsjtxStatus package = do
      status <- fmap join $ tryReadMVar wsjtxStatus
      case package of
        WSJTX.PDecode msg -> do
           let (t,m) =  toMqttMsg mainConfig status msg
           publish mqttConf NoConfirm False t m


        WSJTX.PStatus s -> modifyMVar_ wsjtxStatus (const $ return $ Just s)
        _  -> return ()


src/W2MTypes.hs
----------------------------------------------------------------------------
-- |
-- Module      :  W2MTypes
-- Copyright   :  (c) Marc Fontaine 2017-2018
-- License     :  BSD3
--
-- Maintainer  :  Marc.Fontaine@gmx.de
-- Stability   :  experimental
-- Portability :  GHC-only
--

{-# LANGUAGE DataKinds           #-}
{-# LANGUAGE DeriveGeneric       #-}
{-# LANGUAGE DeriveLift          #-}
{-# LANGUAGE OverloadedStrings   #-}
{-# LANGUAGE RecordWildCards     #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TemplateHaskell     #-}

module W2MTypes where

import           Data.Aeson as Aeson
import qualified Data.ByteString ()
import qualified Data.ByteString.Char8 as BS (pack)
import           Data.Default.Class
import           Data.Fixed
import           Data.Text (Text)
import           Data.Time
import           Data.Word
import           Data.Yaml as Yaml
import           GHC.Generics
import           Instances.TH.Lift ()
import           Language.Haskell.TH.Quote
import           Language.Haskell.TH.Syntax
import           Options hiding (Options, defaultOptions)
import qualified Options

data MainOptions = MainOptions
    { _debug       :: Bool
     ,_configPath  :: Maybe String
     ,_testDefault :: Bool
     }

instance Options.Options MainOptions where
    defineOptions = pure MainOptions
        <*> simpleOption "debug" False
            "enable debugging output"
        <*> simpleOption "config" Nothing
            "path of the configfile"
        <*> simpleOption "testDefault" False
            "just for testing ignore config file and use default config"

data DumpOptions
  = DumpText | DumpHaskell | DumpJSON
  deriving (Show, Eq, Ord, Enum, Bounded)

instance Options.Options DumpOptions where
    defineOptions = defineOption (optionType_enum "DumpHaskell | DumpText | DumpJSON" ) (\o -> o
            { optionLongFlags = ["format"]
            , optionDefault = DumpHaskell
            })

data EmptyOptions = EmptyOptions
instance Options.Options EmptyOptions
  where defineOptions = pure EmptyOptions

data Config = Config {
    config_reporter         :: ReporterConfig
  , config_report_broker    :: BrokerConfig
  , config_enable_report    :: Bool
  , config_wsjtx            :: WSJTXConfig
  , config_protocol_version :: Text
  , config_debug            :: Bool
  } deriving (Show, Read, Eq, Generic, Lift)

instance ToJSON Config where
  toJSON = genericToJSON aesonOptionsDropPrefix
  toEncoding = genericToEncoding aesonOptionsDropPrefix
instance FromJSON Config where
  parseJSON = genericParseJSON aesonOptionsDropPrefix

instance Default Config where
  def = Config {
    config_reporter= def
  , config_report_broker = def
  , config_enable_report = True
  , config_wsjtx = def
  , config_protocol_version ="W2M-V0.1"
  , config_debug = False
  }

data BrokerConfig = BrokerConfig {
   username        :: Maybe Text
  ,password        :: Maybe Text
  ,host            :: String
  ,head_topics     :: [Text]
  ,beacon_interval :: Maybe Int
  ,beacon_topic    :: Text
  } deriving (Show, Read, Eq, Generic, Lift)

instance ToJSON BrokerConfig where
  toJSON = genericToJSON defaultOptions
  toEncoding = genericToEncoding defaultOptions
instance FromJSON BrokerConfig

instance Default BrokerConfig where
  def = BrokerConfig {
   username = Just "myUserName"
  ,password = Just "secret"
  ,host     = "localhost"
--  ,port     =
  ,head_topics = ["hamradio","spot"]
  ,beacon_interval = Nothing
  ,beacon_topic = "test-beacon"
  }

data ReporterConfig = ReporterConfig {
   reporter_callsign    :: Maybe Text
  ,reporter_locator     :: Text
  ,reporter_info        :: Text
  ,reporter_band_config :: BandReport
  } deriving (Show, Read, Eq, Generic, Lift)
instance ToJSON ReporterConfig where
  toJSON = genericToJSON aesonOptionsDropPrefix
  toEncoding = genericToEncoding aesonOptionsDropPrefix
instance FromJSON ReporterConfig where
  parseJSON = genericParseJSON aesonOptionsDropPrefix

instance Default ReporterConfig where
  def = ReporterConfig {
   reporter_callsign = Just "E8TST"
  ,reporter_locator  = "FE43ab"
  ,reporter_info  = "rig : FT-840,  Antenna : Vertical"
  ,reporter_band_config = TrackWSJTXBand
  }

data BandReport
  = UnknownBand
  | FixedBand Text
  | TrackWSJTXBand
  deriving (Show, Read, Eq, Generic, Lift)

instance ToJSON BandReport where
  toJSON = genericToJSON defaultOptions
  toEncoding = genericToEncoding defaultOptions
instance FromJSON BandReport

data WSJTXConfig = WSJTXConfig {
   udp_port  :: Integer
  } deriving (Show, Read, Eq, Generic, Lift)

instance ToJSON WSJTXConfig where
  toJSON = genericToJSON defaultOptions
  toEncoding = genericToEncoding defaultOptions
instance FromJSON WSJTXConfig

instance Default WSJTXConfig where
  def = WSJTXConfig {
   udp_port = 2237
  }

data Report = Report {
   message         :: Text
  ,mode            :: Text
  ,band            :: Text
  ,freq            :: Word64
  ,time            :: DiffTime
  ,snr             :: Int
  ,delta_time      :: Fixed E2
  ,delta_frequency :: Word32
  ,recv_locator    :: Text
  ,recv_callsign   :: Text
  ,recv_info       :: Text
  } deriving (Show, Eq, Generic)

instance ToJSON Report where
  toJSON = genericToJSON defaultOptions
  toEncoding = genericToEncoding defaultOptions
instance FromJSON Report

quoteConfig :: QuasiQuoter
quoteConfig = QuasiQuoter {
   quoteExp = \str -> do
       a <- lift str
       b <- lift $ stringToConfig str
       return $ TupE [a,b]
  ,quotePat = const $ error "No quotePat defined for quoteConfig"
  ,quoteType = const $ error "No quoteType defined for quoteConfig"
  ,quoteDec = const $ error "No quoteDec defined for quoteConfig"
  }

stringToConfig :: String -> Config
stringToConfig str
  = case Yaml.decodeEither' $ BS.pack str of
     Right x  -> x
     Left msg -> error $ ("stringToConfig " ++ show  msg)

aesonOptionsDropPrefix :: Options
aesonOptionsDropPrefix
  = defaultOptions {
      fieldLabelModifier = tail . dropWhile (not . (==) '_')
    }

debugPrint :: Config -> String -> IO ()
debugPrint c s = if config_debug c
  then putStrLn s
  else return ()
